public/app/features/explore/state/reducers.ts

```diff
@@ -0,0 +1,437 @@
+import {
+  calculateResultsFromQueryTransactions,
+  generateEmptyQuery,
+  getIntervals,
+  ensureQueries,
+} from 'app/core/utils/explore';
+import { ExploreItemState, ExploreState, QueryTransaction } from 'app/types/explore';
+import { DataQuery } from 'app/types/series';
+
+import { Action, ActionTypes } from './actions';
+
+export const DEFAULT_RANGE = {
+  from: 'now-6h',
+  to: 'now',
+};
+
+// Millies step for helper bar charts
+const DEFAULT_GRAPH_INTERVAL = 15 * 1000;
+
+/**
+ * Returns a fresh Explore area state
+ */
+const makeExploreItemState = (): ExploreItemState => ({
+  StartPage: undefined,
+  containerWidth: 0,
+  datasourceInstance: null,
+  datasourceError: null,
+  datasourceLoading: null,
+  datasourceMissing: false,
+  exploreDatasources: [],
+  history: [],
+  initialQueries: [],
+  initialized: false,
+  modifiedQueries: [],
+  queryTransactions: [],
+  queryIntervals: { interval: '15s', intervalMs: DEFAULT_GRAPH_INTERVAL },
+  range: DEFAULT_RANGE,
+  scanning: false,
+  scanRange: null,
+  showingGraph: true,
+  showingLogs: true,
+  showingTable: true,
+  supportsGraph: null,
+  supportsLogs: null,
+  supportsTable: null,
+});
+
+/**
+ * Global Explore state that handles multiple Explore areas and the split state
+ */
+const initialExploreState: ExploreState = {
+  split: null,
+  left: makeExploreItemState(),
+  right: makeExploreItemState(),
+};
+
+/**
+ * Reducer for an Explore area, to be used by the global Explore reducer.
+ */
+const itemReducer = (state, action: Action): ExploreItemState => {
```

Comment:
I followed the advice in the [docs](https://redux.js.org/recipes/structuring-reducers/splitting-reducer-logic)
Perhaps you can help me find an obvious slice for more composition?
As for splitting up by grouping thematically, I think this is more confusing IMO and AFAIK I dont think it affects testability.
Also, since reading Ousterhout, I'm less inclined to split things up for organisational reasons, but perhaps I'm overlooking other benefits.

Reply:
Maybe simply defining a simple reducer per switch statement would be sufficient. So, instead of having:

```ts
case ActionTypes.ChangeTime: {
  return {
     ...state,
    range: action.range,
  };
}
```

I would go with:
```
case ActionTypes.ChangeTime: changeTimeReducer(state, actionPayload)
```
 
This way we could have reducers more testable.